Использование аннотации @Query для вызова SQL из приложения.

Manikgog opened the pull request:
Использована аннотация @Query и добавлен следующий функционал в проект:
- Возможность получить количество всех студентов в школе. Эндпоинт должен вернуть число.
- Возможность получить средний возраст студентов. Эндпоинт должен вернуть число.
- Возможность получать только пять последних студентов. Последние студенты считаются теми, у кого идентификатор больше других.

Добавление пагинации к аватаркам.
Добавлена пагинация для репозитория AvatarRepository и контроллер в AvatarTController, чтобы можно было получать списки аватарок постранично.

Manikgog committed:
Использование аннотации @Query для вызова SQL из приложения.
Использована аннотация @Query и добавьлен следующий функционал в проект:
- Возможность получить количество всех студентов в школе. Эндпоинт должен вернуть число.
- Возможность получить средний возраст студентов. Эндпоинт должен вернуть число.
- Возможность получать только пять последних студентов. Последние студенты считаются теми, у кого идентификатор больше других.

Добавление пагинации к аватаркам.
Добавлена пагинация для репозитория AvatarRepository и контроллер в AvatarTController, чтобы можно было получать списки аватарок постранично.

Manikgog committed:
Дописаны методы для тестирования приложения в класс StudentControllerTestRestTemplateTest:
getStudentsAmountTest() для тестирования метода getStudentsAmount();
getAverageAgeTest() для тестирования метода getAverageAge();
get5LastStudentsTest() для тестирования метода get5LastStudents().

dmitry-bizin commented on src/main/java/ru/hogwarts/school/repository/AvatarRepository.java:
```diff
@@ -1,9 +1,14 @@
 
 public interface AvatarRepository extends JpaRepository<Avatar, Long> {
     Optional<Avatar> findById(long studentId);
+    @Query(value = "SELECT data FROM public.avatars ORDER BY id ASC ", nativeQuery = true)
+    List<byte[]> getAvaratrsData();
```
метод вроде бы не используется нигде

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/AvatarService.java:
```diff
@@ -100,4 +103,10 @@ private byte[] generateImagePreview(Path file) throws IOException{
     }
+
+    public ResponseEntity<List<Avatar>> getAvatarsList(int pageNumber, int size) {
+        PageRequest pageRequest = PageRequest.of(pageNumber - 1, size);
```
хорошо бы сделать валидацию входных параметров, иначе можно передать отрицательные, например, и получить ошибку

dmitry-bizin commented on src/test/java/ru/hogwarts/school/controller/StudentControllerTestRestTemplateTest.java:
```diff
@@ -365,4 +365,47 @@ public void getNegativeTest(){
+    @Test
+    public void get5LastStudentsTest(){
+        List<Student> expectedList = new ArrayList<>();
+        for (int i = students.size() - 1; i >= students.size() - 5; i--) {
+            expectedList.add(students.get(i));
+        }
```
а списка есть метод subList, попробуй)

Manikgog committed:
Исправлены замечания. В методе get5LastStudentsTest() цикл для получения пяти последних студентов заменён на метод subList(from, to).
Для валидации переданных параметров в метод getListAvatar(@RequestParam int pageNumber, @RequestParam int size) AvatarController в классе АvatarService добавлен метод validate(int pageNumber, int size), который проверяет параметры и выбрасывает исключения в случае если они отрицательные.
Из интерфейса AvatarRepository удален метод.

Manikgog committed:
Дописаны методы для тестирования приложения в класс StudentControllerTestRestTemplateTest:
getStudentsAmountTest() для тестирования метода getStudentsAmount();
getAverageAgeTest() для тестирования метода getAverageAge();
get5LastStudentsTest() для тестирования метода get5LastStudents().
